handling 'a' tag to map as clickable external url in the generated do…

ahmedelmahalawey opened the pull request:
@senny & @stadelmanma 
I have implemented 'a' tag handling to be mapped as a clickable url in the generated docx e.g. `<a href="http://www.google.com">google</a>` will be interpreted as [google](http://www.google.com) in the generated docx.

Unfortunately I have a very tight schedule so I didn't have anytime to write tests for this feature, so please check it and lets see how can we merge it to the master.

Thanks

amrfaisal commented:
@senny and @stadelmanma : Do you think we could have this reviewed to be merged anytime soon?

stadelmanma commented:
Hi @ahmedelmahalawey thank you for your work on this, hyperlink support will be a great future addition. I have a couple of  questions/comments regarding things as they stand. 

We do require new features to be covered by tests. Our suite is relatively light tending towards the integration side of things so they are quick to write (I'm short on time myself so I understand). You can see unit test examples in `test/html` (you would only need a few assertions in there). As for integration tests we would want one that shows a valid doc can be generated when the `document.xml.rels` file is missing as well as updating the HTML insertion test fixture.

The AST class looks  good it just needs a couple of tweaks light tweaks. I like the use of SecureRandom to ensure a unique ID. As far as I know MS Word won't complain about letters in the `rId`.
   1. The `{ rStyle: 'HyperLink' }` bit should be defined on the Config object (like the `<p>` tag does with `pStyle`) instead of being the only property passed down to the child nodes.  
   2. The `transferred_properties` method should be used to send properties to the child nodes (check out the Paragraph AST class)
   3. The accessor for `@runs` could be dropped, the way AST classes are used nothing should need to access that type of attribute

In the `relationships.rb` file, we would rather avoid singletons. Instead an instance should be registered to the environment that gets passed around (the `env` argument in AST classes) and interacted with in that way. It's a little more of a pain since the `env` gets "cloned" in some cases because the context has to change but it avoids the need for singletons. Essentially, follow the same pattern as `lib/sablon/numbering.rb`'s interaction with `env`.

Have you tried running this on a document without a `document.xml.rels` file? I"m not sure if this file is ever not present but it is something I'll check on myself as well.

Thank you again for your work on this PR! I hope we can get it merged soon. 
- Matt

ahmedelmahalawey commented:
@stadelmanma Thanks for your reply, I will work on your notes very soon and hopefully we merge it could be merged then.

Thanks

ahmedelmahalawey commented:
Hi @stadelmanma, 

1. I have implemented 1 test function called test_a in text/html/ast_tests.rb file. Also I tried to add a hyperlink to the sample document and added a tage with the same value in the html_test_content.html file to check the resulting file but it failed dues to the value of the r:id set in the generated file will never be equal to the one in the sample file and testing this part will need alot of changes in the following function 'assert_docx_equal'.

2. I have moved the `{rStyle: Hyperlink}` to the configuration file.

3. I have used the `transferred_properties` to pass the `{rStyle: Hyperlink}` to the hyperlink child.

4- I don't understand the problem in using @run, in fact I used this behavior in reference to the paragraph Ast Class. I used @run as I needed a global variable in the hyperlink class to build the node's children as Hyperlink is more of a node type than a run type. I really don't understand the issue in this point and need more clarification from you.

5. I have created a relationship instance in the environment class as advised to avoid using singleton class.

Thanks

stadelmanma commented on test/html_test.rb:
```diff
@@ -44,6 +44,6 @@ def content
     html_str = html_str.gsub(/\s+/, ' ')
     # clear any white space between block level tags and other content
-    html_str.gsub(%r{\s*<(/?(?:h\d|div|p|br|ul|ol|li).*?)>\s*}, '<\1>')
+    html_str.gsub(%r{\s*<(/?(?:h\d|div|p|br|ul|ol|li|a).*?)>\s*}, '<\1>')
```
The anchor tag isn't block level, so we want to preserve it's surrounding whitespace otherwise well get stuff like "textwww.google.comtext" instead of the desired "text  www.google.com  text".

Also, we will need to add some links to the `html_test_content` fixture once the testing issue with SecureRandom is squared up. Preferably one without any styles and one with some basic formatting. This integration test is probably our most important test since it allows us to see how MS Word will actually interpret things. Especially if/when we find a case where Microsoft deviated from the formal OOXML spec.

stadelmanma commented on test/html/ast_test.rb:
```diff
@@ -32,6 +33,12 @@ def test_i
+  def test_a
+    input = '<p>Lorem <a href="http://www.google.com">google</a></p>'
+    ast = @converter.processed_ast(input)
+    assert_equal '<Root: [<Paragraph{Paragraph}: [<Run{}: Lorem >, <Hyperlink{}: [<Run{rStyle=Hyperlink}: google>]>]>]>', ast.inspect
```
This test is good, although it makes me wonder if the inspect output of the HyperLink AST node should include the link URL. 

Also, we will need a matching assertion in `test/html/converter_test.rb` that checks the XML generated.

ahmedelmahalawey replied:
1- I have added the target url to the inspect string, it will be appended to the set of propeties
2- I have created 2 tests in the HTMLConverterTest class where 'a' tag is inside 'div' and another where it is inside 'p', and 1 test in HTMLConverterStyleTest where text text inside 'a' tag is bold or italic

stadelmanma commented on lib/sablon.rb:
```diff
@@ -3,6 +3,7 @@
 
 require "sablon/version"
 require "sablon/configuration/configuration"
+require "sablon/relationship/relationship"
```
I'm not sure if this file really needs nested in a directory but it doesn't hurt anything either.

ahmedelmahalawey replied:
I agree with you, I will move the relationship.rb file to the sablon directory

stadelmanma commented:
Hi @ahmedelmahalawey 

Re. #4 I apologize for not being clear, I wasn't meaning the attribute `@runs` itself that portion is perfectly fine. As you noted child elements need somewhere to be stored. I was meaning the actual `attr_accessor :runs` on line 397.  Technically we don't need a public getter and setter for that attribute with the current logic but thats  a minor thing since having one doesn't hurt either (unless I missed something).

Everything looks pretty good, I added a few review comments. Only minor adjustments and I think we should be good to go soon. I'm currently working on permitting inline HTML insertion as time allows which will complement your work very nicely (ref #48).

stadelmanma commented on test/test_helper.rb:
```diff
@@ -14,4 +14,20 @@ class Sablon::TestCase < MiniTest::Test
+
+  class UIDTestGenerator
+    def initialize
+      @current_id = rand(10...100)
```
This was a good solution to the random ID generation. Although you might want to fix the initial ID value (i.e. like  `@current_id = 1234`) otherwise the html insertion integration test will still fail even though we have a predictable next uuid value because we can't modify the "expected" word ml prior to the equality assertion like you do in converter_test.rb.

ahmedelmahalawey replied:
Yeah I just did that and tried integration test but there was a problem that I need your opinion for, I tried to create a hyperlink using microsoft office 2016 in html_sample.docx file so I can compare it with the generated file, but apparently it adds more attributes and tags to the file document.xml that is not added by sablon generator, my guess is new features supported by office 2016. So I unzipped the html_sample.docx and added the hyperlinks manually, I am not sure if this is a valid way to test it, so I need your opinion in that matter.

Thanks

stadelmanma replied:
The way I usually refresh the sample document is generate it straight from sablon from the template. Once the output looks and works like I expect it to then I commit the new docx sample and push. Occasionally, I'll unzip it and double check the contents if I want to be extra sure.

I don't actually open and edit the sample doc because like you said MS Word will pack a bunch of extra junk in there such as save and edit information into the XML.

ahmedelmahalawey replied:
Well in this case I have generated a file with sablon and double checked it, it contains three hyperlinks; one normal, one bold and one italic then I replaced the html_sample.docx file with it. I think I have finished all of the reviews please check again and if there is no other issues I guess we are ready for the merge.

Thanks

stadelmanma commented:
@ahmedelmahalawey everything is good to go. Thank you for all of your work! This is a great addition to Sablon and helps lower the barrier for other features that require modifications to the document.xml.rels file.

amrfaisal commented:
Great news, it's a great addition indeed. Could we also push a new gem version?

stadelmanma commented:
Hi @amrfaisal, I am going to merge the HTML Table support branch and implement inline HTML insertion prior to tagging a new release (ideally this will be soon). However, you shouldn't have any issues pulling the gem straight from the github repo (master branch) in the meantime unless your workflow prevents it.

amrfaisal commented:
@stadelmanma, No worries, we can get it directly from github for sure, and it make a lot of sense to wait for other HTML support first, as this will reflect as a great step for this milestone #47 
Thanks a lot for all the great work guys.

stadelmanma commented:
@ahmedelmahalawey and @amrfaisal i have tagged a new release for sablon with all of the changes mentioned in #47 to extend current HTML insertion support, some of the early tasks were included in the prior release.

amrfaisal commented:
Thanks a lot @stadelmanma this is a great milestone for Sablon.
